lib/Language/Haskell/Stylish/GHC.hs

```diff
@@ -0,0 +1,99 @@
+import           FileSettings                    (FileSettings(..))
+import           GHC.Fingerprint                 (fingerprint0)
+import           GHC.Platform
+import           GHC.Version                     (cProjectVersion)
+import           GhcNameVersion                  (GhcNameVersion(..))
+import           PlatformConstants               (PlatformConstants(..))
+import           SrcLoc                          (GenLocated(..), SrcSpan(..))
+import           SrcLoc                          (Located, RealLocated)
+import           SrcLoc                          (srcSpanStartLine, srcSpanEndLine)
+import           ToolSettings                    (ToolSettings(..))
+import qualified Outputable                      as GHC
+
+getStartLineUnsafe :: Located a -> Int
+getStartLineUnsafe = \case
+  (L (RealSrcSpan s) _) -> srcSpanStartLine s
+  _ -> error "could not get start line of block"
+
+getEndLineUnsafe :: Located a -> Int
+getEndLineUnsafe = \case
+  (L (RealSrcSpan s) _) -> srcSpanEndLine s
+  _ -> error "could not get end line of block"
+
+dropAfterLocated :: Maybe (Located a) -> [RealLocated b] -> [RealLocated b]
+dropAfterLocated loc xs = case loc of
+  Just (L (RealSrcSpan rloc) _) ->
+    filter (\(L x _) -> srcSpanEndLine rloc >= srcSpanStartLine x) xs
+  _ -> xs
+
+dropBeforeLocated :: Maybe (Located a) -> [RealLocated b] -> [RealLocated b]
+dropBeforeLocated loc xs = case loc of
+  Just (L (RealSrcSpan rloc) _) ->
+    filter (\(L x _) -> srcSpanStartLine rloc <= srcSpanEndLine x) xs
+  _ -> xs
+
+dropBeforeAndAfter :: Located a -> [RealLocated b] -> [RealLocated b]
+dropBeforeAndAfter loc = dropBeforeLocated (Just loc) . dropAfterLocated (Just loc)
+
+baseDynFlags :: GHC.DynFlags
+baseDynFlags = defaultDynFlags fakeSettings llvmConfig
+  where
+    fakeSettings = GHC.Settings
+      { sGhcNameVersion = GhcNameVersion "stylish-haskell" cProjectVersion
+      , sFileSettings = FileSettings {}
+      , sToolSettings = ToolSettings
+        { toolSettings_opt_P_fingerprint = fingerprint0,
+          toolSettings_pgm_F = ""
+        }
+      , sPlatformConstants = PlatformConstants
+        { pc_DYNAMIC_BY_DEFAULT = False
+        , pc_WORD_SIZE = 8
+        }
+      , sTargetPlatform = Platform
+        { platformMini = PlatformMini
+          { platformMini_arch = ArchUnknown
+          , platformMini_os = OSUnknown
+          }
+        , platformWordSize = PW8
+        , platformUnregisterised = True
+        , platformHasIdentDirective = False
+        , platformHasSubsectionsViaSymbols = False
+        , platformIsCrossCompiling = False
+        }
+      , sPlatformMisc = PlatformMisc {}
+      , sRawSettings = []
+      }
+
+    llvmConfig = GHC.LlvmConfig [] []
+
+unLocated :: Located a -> a
+unLocated (L _ a) = a
+
+showOutputable :: GHC.Outputable a => a -> String
+showOutputable = GHC.showPpr baseDynFlags
+
+compareOutputable :: GHC.Outputable a => a -> a -> Ordering
```

Comment:
Do we have any other orderings in use? If not maybe it would make sense to extract this to an orphan `Ord` instance for `GHC.Outputable` and not use this one explicitly?

Reply:
This is mostly used in the imports step. I think it might be a good idea here to invent our type for imports that carries the information we need. This'll probably be nice when we add support for merging imports - so then we add the ordering instance for that type.

WDYT?

Reply:
Sounds good!

Reply:
So I did two things, I changed the newtype representation of:

```haskell
newtype Imports = Imports [LImportDecl GhcPs]

-- to:

newtype Import = Import (ImportDecl GhcPs)
```

This makes things a bit easier to reason about, it was possible to define equality for the import individually and I think it'll give us a more intuitive API down the line to work with imports. E.g. I added a helper function `canMergeImport` which is used in the equality instance. This helper could be used in the `Imports` step to merge imports - and get better ordering... I think I'm gonna play with that a bit today

Reply:
@lukasz-golebiewski @felixmulder

> Do we have any other orderings in use? If not maybe it would make sense to extract this to an orphan `Ord` instance for `GHC.Outputable` and not use this one explicitly?

This thread caught my eye as I perused this ticket.  It looks to me that these lines are covering the same ground as [`ghc-lib-parser-ex`](https://hackage.haskell.org/package/ghc-lib-parser-ex)'s [`HsExtendInstances` wrapper](https://hackage.haskell.org/package/ghc-lib-parser-ex-8.10.0.16/docs/Language-Haskell-GhclibParserEx-GHC-Hs-ExtendInstances.html). I've [mused before](https://github.com/ndmitchell/hlint/issues/1004) that `ghc-lib-parser-ex` might be useful for stylish-haskell to use/extend on its conversion from haskell-src-exts to `ghc-lib-parser` - would be nice to share these sorts of common utilities with e.g. HLint!

Reply:
Hmm c.f. `showOutputable :: GHC.Outputable a => a -> String` with [`unsafePrettyPrint`](https://hackage.haskell.org/package/ghc-lib-parser-ex-8.10.0.16/docs/Language-Haskell-GhclibParserEx-GHC-Utils-Outputable.html#v:unsafePrettyPrint) too.